File: src/marketschema/adapters/registry.py
"""Adapter registry for managing multiple data source adapters."""

from marketschema.adapters.base import BaseAdapter
from marketschema.exceptions import AdapterError


class AdapterRegistry:
    """Registry for managing adapter instances by source name.

    This is a singleton registry that allows adapters to be registered
    and retrieved by their source_name.

    Example:
        @register
        class BinanceAdapter(BaseAdapter):
            source_name = "binance"
            ...

        # Get adapter by name
        adapter = AdapterRegistry.get("binance")
    """

    _instance: "AdapterRegistry | None" = None
    _adapters: dict[str, type[BaseAdapter]]

    def __new__(cls) -> "AdapterRegistry":
        """Create singleton instance."""
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._adapters = {}
        return cls._instance

    @classmethod
    def register[T: BaseAdapter](cls, adapter_class: type[T]) -> type[T]:
        """Register an adapter class with the registry.

        Args:
            adapter_class: Adapter class to register

        Returns:
            The adapter class (for use as decorator)

        Raises:
            AdapterError: If adapter has no source_name or name is already registered
        """
        instance = cls()

        source_name = adapter_class.source_name
        if not source_name:
            raise AdapterError(
                f"{adapter_class.__name__} must define source_name"
            )

        if source_name in instance._adapters:
            raise AdapterError(
                f"Adapter for '{source_name}' is already registered"
            )

        instance._adapters[source_name] = adapter_class
        return adapter_class

    @classmethod
    def get(cls, source_name: str) -> BaseAdapter:
        """Get an adapter instance by source name.

        Args:
            source_name: Name of the data source

        Returns:
            Instance of the registered adapter

        Raises:
            KeyError: If no adapter is registered for the source name
        """
        instance = cls()

        if source_name not in instance._adapters:
            available = ", ".join(instance._adapters.keys()) or "none"
            raise KeyError(
                f"No adapter registered for '{source_name}'. "
                f"Available adapters: {available}"
            )

        adapter_class = instance._adapters[source_name]
        return adapter_class()

    @classmethod
    def list_adapters(cls) -> list[str]:
        """List all registered adapter source names.

        Returns:
            List of registered source names
        """
        instance = cls()
        return list(instance._adapters.keys())

    @classmethod
    def clear(cls) -> None:
        """Clear all registered adapters.

        Primarily useful for testing.
        """
        instance = cls()
        instance._adapters.clear()

    @classmethod
    def is_registered(cls, source_name: str) -> bool:
        """Check if an adapter is registered for a source name.

        Args:
            source_name: Name of the data source

        Returns:
            True if an adapter is registered, False otherwise
        """
        instance = cls()
        return source_name in instance._adapters


def register[T: BaseAdapter](adapter_class: type[T]) -> type[T]:
    """Decorator to register an adapter class with the global registry.

    Example:
        @register
        class MyAdapter(BaseAdapter):
            source_name = "my_source"
            ...

    Args:
        adapter_class: Adapter class to register

    Returns:
        The adapter class
    """
    return AdapterRegistry.register(adapter_class)


__all__ = ["AdapterRegistry", "register"]

File: crates/marketschema-adapters/src/registry.rs
//! Thread-safe global adapter registry.

use crate::adapter::BaseAdapter;
use crate::error::AdapterError;
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Type alias for adapter factory functions.
pub type AdapterFactory = Arc<dyn Fn() -> Box<dyn BaseAdapter> + Send + Sync>;

/// Global registry for adapters.
static REGISTRY: Lazy<RwLock<HashMap<String, AdapterFactory>>> =
    Lazy::new(|| RwLock::new(HashMap::new()));

/// Thread-safe global registry for adapter discovery.
///
/// The registry allows adapters to be registered by their source name
/// and retrieved later for dynamic adapter selection.
///
/// # Example
///
/// ```ignore
/// use marketschema_adapters::AdapterRegistry;
///
/// // Register an adapter
/// AdapterRegistry::register("myapi", || Box::new(MyAdapter::new()))?;
///
/// // Get an adapter instance
/// if let Some(adapter) = AdapterRegistry::get("myapi")? {
///     println!("Source: {}", adapter.source_name());
/// }
/// ```
pub struct AdapterRegistry;

impl AdapterRegistry {
    /// Registers an adapter factory with the given source name.
    ///
    /// Returns an error if an adapter with the same name is already registered,
    /// or if the registry lock is poisoned.
    pub fn register<F>(source_name: impl Into<String>, factory: F) -> Result<(), AdapterError>
    where
        F: Fn() -> Box<dyn BaseAdapter> + Send + Sync + 'static,
    {
        let name = source_name.into();
        let mut registry = REGISTRY.write().map_err(|e| {
            AdapterError::LockPoisoned(format!("Failed to acquire write lock: {}", e))
        })?;

        if registry.contains_key(&name) {
            return Err(AdapterError::DuplicateRegistration(name));
        }

        registry.insert(name, Arc::new(factory));
        Ok(())
    }

    /// Gets an adapter instance by source name.
    ///
    /// Returns `Ok(None)` if no adapter is registered with the given name.
    /// Returns an error if the registry lock is poisoned.
    pub fn get(source_name: &str) -> Result<Option<Box<dyn BaseAdapter>>, AdapterError> {
        let registry = REGISTRY.read().map_err(|e| {
            AdapterError::LockPoisoned(format!("Failed to acquire read lock: {}", e))
        })?;
        Ok(registry.get(source_name).map(|factory| factory()))
    }

    /// Returns a list of all registered adapter source names.
    ///
    /// Returns an error if the registry lock is poisoned.
    pub fn list_adapters() -> Result<Vec<String>, AdapterError> {
        let registry = REGISTRY.read().map_err(|e| {
            AdapterError::LockPoisoned(format!("Failed to acquire read lock: {}", e))
        })?;
        Ok(registry.keys().cloned().collect())
    }

    /// Checks if an adapter is registered with the given source name.
    ///
    /// Returns an error if the registry lock is poisoned.
    pub fn is_registered(source_name: &str) -> Result<bool, AdapterError> {
        let registry = REGISTRY.read().map_err(|e| {
            AdapterError::LockPoisoned(format!("Failed to acquire read lock: {}", e))
        })?;
        Ok(registry.contains_key(source_name))
    }

    /// Clears all registered adapters.
    ///
    /// This is intended for testing to ensure test isolation.
    ///
    /// # Warning
    ///
    /// Do not use in production code as it removes all registered adapters.
    #[cfg(any(test, feature = "test-utils"))]
    pub fn clear() -> Result<(), AdapterError> {
        let mut registry = REGISTRY.write().map_err(|e| {
            AdapterError::LockPoisoned(format!("Failed to acquire write lock: {}", e))
        })?;
        registry.clear();
        Ok(())
    }
}
